/**
 * Copyright (c) 2015 - The CM Authors <legal@clickmatcher.com>
 *   All Rights Reserved.
 *
 * This file is CONFIDENTIAL -- Distribution or duplication of this material or
 * the information contained herein is strictly forbidden unless prior written
 * permission is obtained.
 */
package zbase;

/**
 * Describes the ACL policy used for a document.
 *
 *   - DOCACL_DEFAULT_PRIVATE
 *     The document is accessible only to users in explicitly mentioned in the
 *     document ACLs
 *
 *   - DOCACL_DEFAULT_PUBLIC_ORGANIZATION
 *     The document is accessible to all users within the organization
 *
 *   - DOCACL_DEFAULT_PUBLIC_INTERNET
 *     The document is accessible to all internet users that have the link
 *
 */
enum DocumentACLPolicy {
  ACLPOLICY_PRIVATE = 1;
  ACLPOLICY_PUBLIC_ORGANIZATION = 2;
  ACLPOLICY_PUBLIC_INTERNET = 3;
};

enum DocumentPublishingStatus {
  PUBSTATUS_DRAFT = 1;
  PUBSTATUS_PUBLISHED = 2;
};

message DocumentACL {

  required string userid = 1;

  optional bool allow_read = 2;

  optional bool allow_write = 3;

  optional bool allow_share = 4;

  optional bool is_owner = 5;

};

message DocumentReference {

  required string uuid = 1;

};

message Document {

  required string uuid = 1;

  required uint64 version = 2;

  required string name = 3;

  optional string type = 4;

  optional uint64 ctime = 5;

  optional uint64 mtime = 6;

  optional uint64 atime = 7;

  optional bool deleted = 14;

  required DocumentACLPolicy acl_policy = 8;

  repeated DocumentACL acls = 9;

  optional DocumentPublishingStatus publishing_status = 12;

  optional string category = 13;

  optional bytes content = 11;

}


